library(shiny)
library(SingleR)
library(celldex)
library(hdf5r)
library(Seurat)
library(tidyverse)
library(pheatmap)
library(ggplot2)
library(shinydashboard)
source("Global.R")
ui <- dashboardPage(
  skin = "blue",
  dashboardHeader(
    title = tags$span(
      style = "font-weight: bold; color: white;", 
      "FPCAM"
    ),
    titleWidth = 300
  ),
  dashboardSidebar(
    width = 250,
    sidebarMenu(
      id = 'tab',
      useShinyjs(),
      menuItem(
        "Home", 
        tabName = "home", 
        icon = icon("home"),
        selected = TRUE
      ),
      menuItem(
        "Cell Annotation", 
        tabName = "input", 
        icon = icon("microscope")
      ),
      menuItem(
        "UMAP Visualization", 
        tabName = "umap_", 
        icon = icon("eye")
      ),
      menuItem(
        "Score Heatmap", 
        tabName = "score_", 
        icon = icon("dna")
      ),
      menuItem(
        "Delta Distribution", 
        tabName = "delta_", 
        icon = icon("chart-bar")
      ),
      menuItem(
        "About", 
        tabName = "about", 
        icon = icon("info-circle")
      ),
      # æ¡ä»¶é¢æ¿ï¼Œç”¨äºæ–‡ä»¶ä¸Šä¼ åŒºåŸŸ
      conditionalPanel(
        condition = "input.tab == 'input'",
        div(
          style = "padding: 15px;",
          h4("File Upload"),
          fileInput(
            "file1", 
            label = tags$span(icon("file-excel"), "Upload File 1 (Marker exp)"),
            multiple = FALSE, 
            accept = c('.xlsx', '.xls'),
            width = "100%",
            placeholder = "Choose Excel file"
          ),
          fileInput(
            "file2", 
            label = tags$span(icon("file-excel"), "Upload File 2 (Dictionary)"),
            multiple = FALSE, 
            accept = c('.xlsx', '.xls'),
            width = "100%",
            placeholder = "Choose Excel file"
          ),
          div(
            style = "display: flex; flex-direction: column; gap: 10px; padding-top: 20px;",
            actionButton(
              "run", 
              "Run", 
              icon = icon("play"), 
              class = "btn-success btn-lg",
              style = "width: 75%; padding: 12px; border-radius: 5px;"
            ),
            actionButton(
              "reset", 
              "Reset", 
              icon = icon("undo"), 
              class = "btn-danger btn-lg",
              style = "width: 75%; padding: 12px; border-radius: 5px;"
            )
          )
        )
      )
    )
  ),
  dashboardBody(
    tabItems(
      # Home Tab
      tabItem(
        tabName = "home",
        box(
          width = 12,
          status = "primary",
          solidHeader = TRUE,
          title = "Welcome",
          tags$div(
            style = "text-align: center; padding: 20px;",
            # æ·»åŠ å›¾ç‰‡
            tags$div(
              style = "display: flex; justify-content: center; align-items: center; margin: 30px 0;",
              tags$img(
                src = "cell_analysis.png",  # å›¾ç‰‡éœ€è¦æ”¾åœ¨ www æ–‡ä»¶å¤¹ä¸­
                width = "100%",
                style = "max-width: 1100px; border-radius: 8px; box-shadow: 0 4px 8px rgba(0,0,0,0.1);"
              )
            ),
            tags$p(
              "Upload your Excel files and perform cell annotation analysis with ease.",
              style = "font-size: 16px; text-align: center;"
            )
          )
        )
      ),
      
      # Cell Annotation Tab
      tabItem(
        tabName = "input",  # Make sure this matches your sidebar menu item
        tabsetPanel(
          id = 'main_tabs',
          tabPanel(
            "Instructions", 
            includeMarkdown("./markdown/instructions.md"),
            tags$hr(),
            tags$div(
              style = "padding: 10px;",
              tags$h4("ğŸ“¥ Download Example Data (Click buttons below):"),
              
              tags$div(
                style = "margin-top: 10px;",
                tags$p("ğŸ”¹ Dictionary Example:"),
                downloadButton(
                  outputId = "download_dictionary",
                  label = "Download Dictionary",
                  style = "margin-top: 10px;"
                )
              ),
              
              tags$div(
                style = "margin-top: 20px;",
                tags$p("ğŸ”¹ Expression Example:"),
                downloadButton(
                  outputId = "download_expression_data",
                  label = "Download Marker Expression",
                  style = "margin-top: 10px;"
                )
              )
            )
          ),
          
          tabPanel(
            "Result",
            tableOutput("resultTable"),
            downloadButton("downloadData", "Download Results")
          )
        )
      ),
      
      # Single Cell Analysis Tab
      tabItem(
        tabName = "umap_",
        fluidRow(
          column(
            12,
            box(
              width = NULL,
              status = "primary",
              title = "Analysis Input",
              fileInput(
                "h5_file", 
                "Choose H5 File",
                accept = c(".h5")
              ),
              actionButton(
                "run_umap_analysis", 
                "Run UMAP Analysis",
                class = "btn-primary",
                style = "width: 100%"
              ),
              # æ˜¾å¼æ·»åŠ ä¸‹è½½æŒ‰é’®
              downloadButton("download_umap", "Download UMAP Plot"),
              hr(),
              helpText("Upload a 10X H5 file to start the analysis")
            )
          )
        ),
        fluidRow(
          column(
            12,
            tabBox(
              width = NULL,
              tabPanel(
                "UMAP Plot",
                plotOutput("umap_plot", height = "600px")
              )
            )
          )
        )
      ),
      tabItem(
        tabName = "score_",
        fluidRow(
          column(
            12,
            box(
              width = NULL,
              status = "primary",
              title = "Analysis Input",
              fileInput(
                "h5_file", 
                "Choose H5 File",
                accept = c(".h5")
              ),
              actionButton(
                "run_heatmap_analysis", 
                "Run Heatmap Analysis",
                class = "btn-primary",
                style = "width: 100%"
              ),
              # æ˜¾å¼æ·»åŠ ä¸‹è½½æŒ‰é’®
              downloadButton("download_heatmap", "Download Heatmap"),
              hr(),
              helpText("Upload a 10X H5 file to start the analysis")
            )
          )
        ),
        fluidRow(
          column(
            12,
            tabBox(
              width = NULL,
              tabPanel(
                "Score Heatmap",
                plotOutput("score_heatmap", height = "600px")
              )
            )
          )
        )
      ),
      tabItem(
        tabName = "delta_",
        fluidRow(
          column(
            12,
            box(
              width = NULL,
              status = "primary",
              title = "Analysis Input",
              fileInput(
                "h5_file", 
                "Choose H5 File",
                accept = c(".h5")
              ),
              actionButton(
                "run_delta_analysis", 
                "Run Delta Analysis",
                class = "btn-primary",
                style = "width: 100%"
              ),
              # æ˜¾å¼æ·»åŠ ä¸‹è½½æŒ‰é’®
              downloadButton("download_delta", "Download Delta Plot"),
              hr(),
              helpText("Upload a 10X H5 file to start the analysis")
            )
          )
        ),
        fluidRow(
          column(
            12,
            tabBox(
              width = NULL,
              tabPanel(
                "Delta Distribution",
                plotOutput("delta_plot", height = "600px")
              )
            )
          )
        )
      ),
      # About Tab æ ‡ç­¾é¡µ
      tabItem(
        tabName = "about",
        fluidRow(
          column(
            12,
            box(
              width = NULL,
              status = "info",
              title = "About This Tool",
              tags$div(
                style = "text-align: center; padding: 20px;",
                tags$h3("FPCAM - Cell Annotation Tool"),
                tags$p(
                  "This tool provides a streamlined approach to annotating cells in single-cell RNA sequencing data. ",
                  style = "font-size: 16px; text-align: center;"
                ),
                tags$p(
                  "Developed by: Guojun Liuã€Yan Shi",
                  style = "font-size: 14px; text-align: center;"
                ),
                tags$p(
                  "Contact: gjliu77@qq.comã€1490261148@qq.com",
                  style = "font-size: 14px; text-align: center;"
                )
              )
            )
          )
        )
      )
    )
  )
)
# Server å‡½æ•°
server <- function(input, output, session) {
  options(shiny.maxRequestSize = 300 * 1024^2)  # è®¾ç½®æœ€å¤§ä¸Šä¼ æ–‡ä»¶å¤§å°
  
  temp_folder <- tempdir()  # å®šä¹‰ä¸´æ—¶æ–‡ä»¶å¤¹è·¯å¾„
  cat("ä¸´æ—¶æ–‡ä»¶å¤¹è·¯å¾„:", temp_folder, "\n")  # æ‰“å°è·¯å¾„ä»¥ä¾¿è°ƒè¯•
  
  # å®šä¹‰å“åº”å¼å˜é‡ä»¥å­˜å‚¨ result_data
  result_data <- reactiveVal(NULL)
  
  # Server logic for the download buttons
    output$download_dictionary <- downloadHandler(
      filename = function() {
        "Dictionary_example.xlsx"
      },
      content = function(file) {
        # Copy the file from www folder to the requested location
        file.copy(
          from = "./www/Dictionary_example.xlsx", 
          to = file,
          overwrite = TRUE
        )
      },
    )
    
    output$download_expression_data <- downloadHandler(
      filename = function() {
        "marker_expression_example.xlsx"
      },
      content = function(file) {
        # Copy the file from www folder to the requested location
        file.copy(
          from = "./www/marker_expression_example.xlsx", 
          to = file,
          overwrite = TRUE
        )
      },
    )
  
  
  
  observeEvent(input$run, {
    req(input$file1, input$file2)  # ç¡®ä¿ä¸¤ä¸ªæ–‡ä»¶éƒ½å·²ä¸Šä¼ 
    
    # æ˜¾ç¤ºè¿›åº¦æ¡
    withProgress(message = 'Processing Files...', value = 0, {
      
      # å°†æ–‡ä»¶ä¿å­˜åˆ°ä¸´æ—¶æ–‡ä»¶å¤¹ä¸­ï¼Œä½¿ç”¨åŸå§‹æ–‡ä»¶å
      file1_path <- file.path(temp_folder, input$file1$name)
      file2_path <- file.path(temp_folder, input$file2$name)
      
      # å¤åˆ¶ä¸Šä¼ çš„æ–‡ä»¶åˆ°ä¸´æ—¶æ–‡ä»¶å¤¹
      file.copy(input$file1$datapath, file1_path, overwrite = TRUE)
      file.copy(input$file2$datapath, file2_path, overwrite = TRUE)
      
      # æ£€æŸ¥æ–‡ä»¶æ˜¯å¦æˆåŠŸå¤åˆ¶
      if (file.exists(file1_path) && file.exists(file2_path)) {
        cat("æ–‡ä»¶å·²æˆåŠŸå¤åˆ¶åˆ°ä¸´æ—¶æ–‡ä»¶å¤¹ã€‚\n")
        print(list.files(temp_folder))  # åˆ—å‡ºä¸´æ—¶æ–‡ä»¶å¤¹ä¸­çš„æ–‡ä»¶
        
        # è®¾ç½®ä¸´æ—¶æ–‡ä»¶å¤¹ä¸ºå·¥ä½œç›®å½•ï¼ˆå¯é€‰ï¼‰
        setwd(temp_folder)
        rm(list = ls())
        Marker <- read_excel(file.path(temp_folder, "Marker_expression.xlsx"), 
                             sheet = 1, col_names = TRUE, col_types = NULL, na = "", skip = 0)
        # åˆå§‹åŒ–ä¸€ä¸ªå‘é‡æ¥å­˜å‚¨æ¯åˆ—çš„æ¯”å€¼
        ratios <- list()
        # å¾ªç¯éå†å¶æ•°ç´¢å¼•çš„åˆ—
        sum(!is.na(Marker[, 3]))
        for (j in 1:sum(!is.na(Marker[, 3]))) {
          print(j)
        }
        for (i in seq(2, ncol(Marker), 2)) {
          # è®¡ç®—åˆ—çš„æ€»å’Œ
          column_sum <- sum(Marker[, i], na.rm = TRUE)
          # è®¡ç®—æ¯ä¸ªå€¼ä¸åˆ—æ€»å’Œçš„æ¯”å€¼
          column_ratios <- Marker[, i] / column_sum
          # å°†æ¯”å€¼å‘é‡æ·»åŠ åˆ°ratiosåˆ—è¡¨ä¸­
          ratios[[i]] <- column_ratios
        }
        
        incProgress(0.1)  
        # ### æ‰‹åŠ¨éªŒè¯
        # col_name <- paste0("cluster","_", i/2-1, "_", "Markergene")
        # gene <- Marker[2,col_name]
        # column_sum <- sum(Marker[, 58], na.rm = TRUE)
        # sum(Marker$cluster_28_Express)
        # column_ratios <- Marker[, 58] / column_sum
        ###################################################################
        ## Script name:  è®¡ç®—åˆ—å’ŒåŸºå› åˆ—æ•´åˆ
        ###################################################################
        for (i in seq(1, ncol(Marker), 2)) {
          print(i)
          for (j in seq(2, ncol(Marker), 2)) {
            Express_var <- paste0("cluster","_", j/2-1, "_", "Express")
            # eval(Express_var)
            ratios[[j]][[eval(Express_var)]]
            # print(paste(i,j))
            # Marker[,i]
          }
        }
        incProgress(0.2)
        # Marker[,1]
        # Express_var <- paste0("cluster","_", 4/2-1, "_", "Express")
        # eval(Express_var)
        # ratios[[4]][[eval(Express_var)]]
        # combined_list <- list(Marker[,1], ratios[[2]][[eval(Express_var)]])
        # combined_list <- list(Marker[,3], ratios[[4]][[eval(Express_var)]])
        # combined_list <- do.call(c, list(Marker[,1], ratios[[2]][[eval(Express_var)]]))
        # åˆå§‹åŒ–ä¸€ä¸ªç©ºåˆ—è¡¨
        inner_list <- list()
        # å¤–å±‚å¾ªç¯
        for (i in seq(0, ncol(Marker)/2, 2)) {
          # åˆå§‹åŒ–å†…å±‚åˆ—è¡¨
          inner_list <- list()
          # å†…å±‚å¾ªç¯
          for (j in seq(2, ncol(Marker), 2)) {
            # ä¸ºå½“å‰ä½ç½®ç”Ÿæˆä¸€ä¸ªå€¼
            Express_var <- paste0("cluster","_", j/2-1, "_", "Express")
            eval(Express_var)
            var <- list(ratios[[j]][[eval(Express_var)]])
            combined_list <- list(Marker[,j-1], ratios[[j]][[eval(Express_var)]])
            # combined_list <- paste("Item", i, "Subitem", j, sep = "_")
            # å°†å€¼æ·»åŠ åˆ°å†…å±‚åˆ—è¡¨
            inner_list[[j]] <- combined_list
          }
          # å°†å†…å±‚åˆ—è¡¨æ·»åŠ åˆ°å¤šç»´åˆ—è¡¨ä¸­
          # multi_dimensional_list[[i]] <- inner_list
        }
        
        incProgress(0.3)  
        # æµ‹è¯•ä»åˆ—è¡¨æ‰“å°åŸºå› å’Œè¡¨è¾¾é‡
        # print(inner_list)
        # inner_list[[2]][[1]][[1]][1]
        # inner_list[[2]][[1]][[1]]
        # # count the length of
        # inner_list[[2]][[2]][1]
        # inner_list[[58]][[1]][[1]][1]
        # inner_list[[58]][[2]][1]
        ###################################################################
        ## Script name: æ¯”å¯¹
        ###################################################################
        db <- read_excel(file.path(temp_folder, "Dictionary.xlsx"), 
                         sheet = 1, col_names = TRUE, col_types = NULL, na = "", skip = 0)
        a <- unlist(unique(db$cell_cluster))
        # å…ˆåˆ›å»ºåˆ—å†åˆ›å»ºè¡Œ
        for (j in seq(2, ncol(Marker), 2)) {
          num =j/2-1
          tmp <- paste("Cluster","_",num,sep="")
          db[tmp] <- 0
          print(paste("Calculating",tmp))
          # j = 56
          my_list <- inner_list[[j]][[1]][[1]]
          for(m in 1:nrow(db)){
            # db[m,tmp] <- NA
            for (l in 1:length(na.omit(unlist(my_list)))) { # Calculate the length of list
              # print("The row",m)
              if (toupper(inner_list[[j]][[1]][[1]][l]) == toupper(db$marker_gene[m])) {
                db[m,tmp] <- inner_list[[j]][[2]][l]*db$percent[m]
              }
            }
          }
        }
        write.csv(db, file = file.path(temp_folder, "Cluster.csv"), row.names = FALSE)
        incProgress(0.4)  
        ###################################################################
        ## Script name: Calculate the cellular origin of clusters.
        ###################################################################
        result <- db %>%
          group_by(cell_cluster) %>%
          summarise(Total=sum(Cluster_1)) # set Cluster_0
        tmp <- "Cluster_1" #
        max_index <- which.max(result$Total)
        a<- result[max_index,1]
        print(paste0("The cellular origin of ",tmp," is ",a))
        
        ###################################################################
        ## Script name: 
        ###################################################################
        for (i in seq(2, ncol(Marker), 2)) {
          num =i/2-1
          num_2 = num+2
          tmp <- paste("Cluster",num,sep="_")
          # db[,tmp] <- as.numeric(db[,tmp])
          result <- db %>%
            group_by(cell_cluster) %>%
            summarise(across(starts_with("Cluster"),~sum(.,na.rm = TRUE)))
          max_index <- which.max(result[[num_2]])
          a<- result[max_index,1]
          print(paste0("The cellular origin of ",tmp," is ",a))
        }
        write.csv(result, file = file.path(temp_folder, "result.csv"), row.names = FALSE, quote = TRUE)
        incProgress(0.5)  
        ###################################################################
        ## Script name: æŒ‘å‡ºæœ€é«˜å’Œæ¬¡é«˜å€¼
        ###################################################################
        # è¯»å–CSVæ–‡ä»¶
        data <- read.csv(file.path(temp_folder, "result.csv"))
        # åˆå§‹åŒ–ä¸€ä¸ªç©ºçš„æ•°æ®æ¡†ï¼Œç”¨äºå­˜å‚¨æ¯ä¸€åˆ—çš„æœ€å¤§å€¼ã€æ¬¡é«˜å€¼åŠå¯¹åº”çš„ç»†èƒç¾¤
        result <- data.frame(Cluster = character(), Max_Cell_Cluster = character(), Max_Value = numeric(),
                             Second_Max_Cell_Cluster = character(), Second_Max_Value = numeric(), stringsAsFactors = FALSE)
        # éå†æ¯ä¸€åˆ—ï¼ˆä»ç¬¬äºŒåˆ—åˆ°æœ€åä¸€åˆ—ï¼‰ï¼Œæ‰¾åˆ°æ¯åˆ—çš„æœ€å¤§å€¼ã€æ¬¡é«˜å€¼åŠå¯¹åº”çš„ç»†èƒç¾¤
        for (j in 2:ncol(data)) {
          # æ‰¾åˆ°å½“å‰åˆ—çš„æœ€å¤§å€¼
          max_value <- max(data[, j])
          # æ‰¾åˆ°æœ€å¤§å€¼å¯¹åº”çš„ç»†èƒç¾¤
          max_cell_cluster <- data$cell_cluster[which.max(data[, j])]
          # æ‰¾åˆ°æ¬¡é«˜å€¼ï¼ˆå»é™¤æœ€å¤§å€¼åå†æ‰¾ï¼‰
          temp <- data[, j]
          temp[which.max(temp)] <- -Inf  # å°†æœ€å¤§å€¼æ›¿æ¢ä¸ºè´Ÿæ— ç©·å¤§ï¼Œç¡®ä¿ä¸ä¼šå†æ¬¡è¢«é€‰ä¸­
          second_max_value <- max(temp)
          # æ‰¾åˆ°æ¬¡é«˜å€¼å¯¹åº”çš„ç»†èƒç¾¤
          second_max_cell_cluster <- data$cell_cluster[which.max(temp)]
          # å°†ç°‡åã€æœ€å¤§å€¼å¯¹åº”çš„ç»†èƒç¾¤ã€æ¬¡é«˜å€¼åŠå…¶å¯¹åº”çš„ç»†èƒç¾¤å­˜å…¥ç»“æœæ•°æ®æ¡†
          result <- rbind(result, data.frame(Cluster = names(data)[j], 
                                             Max_Cell_Cluster = max_cell_cluster, Max_Value = max_value, 
                                             Second_Max_Cell_Cluster = second_max_cell_cluster, Second_Max_Value = second_max_value))
        }
        # å°†ç»“æœå†™å…¥CSVæ–‡ä»¶
        write.csv(result, file = file.path(temp_folder, "result_max_second_values_per_column.csv"), row.names = FALSE)
        incProgress(0.6)  
        ###################################################################
        ## Script name: è®¡ç®—deltaåˆ¶è¡¨
        ###################################################################
        # è¯»å–CSVæ–‡ä»¶
        data <- read.csv(file.path(temp_folder, "result.csv"))
        # è¯»å–result_max_second_values_per_column.csvæ–‡ä»¶
        max_data<- read.csv(file.path(temp_folder, "result_max_second_values_per_column.csv"), 
                            header = TRUE, stringsAsFactors = FALSE)
        # åˆ›å»ºä¸€ä¸ªå‘é‡ç”¨äºå­˜å‚¨æ¯åˆ—çš„ delta1 å€¼ï¼Œé•¿åº¦ä¸ºæ•°æ®åˆ—æ•°å‡1(å»æ‰ç¬¬ä¸€åˆ—)
        delta1_values <- numeric(ncol(data) - 1)
        # éå†æ•°æ®æ¡†ä¸­é™¤ç¬¬ä¸€åˆ—å¤–çš„æ‰€æœ‰åˆ—
        for (i in 2:ncol(data)) {
          # æå–å½“å‰åˆ—çš„æ•°æ®ï¼Œdrop=FALSE ä¿æŒæ•°æ®æ¡†ç»“æ„
          column_data <- data[, i, drop = FALSE]
          # è¿‡æ»¤æ•°æ®ï¼š
          column_data <- column_data[!is.na(column_data) & column_data != max(column_data, na.rm = TRUE)]
          # æ£€æŸ¥è¿‡æ»¤åçš„æ•°æ®æ˜¯å¦è¿˜æœ‰å€¼
          if (length(column_data) > 0) {
            # è®¡ç®—è¿‡æ»¤åæ•°æ®çš„ä¸­ä½æ•°
            median_value <- median(column_data, na.rm = TRUE)
            # è®¡ç®— delta1ï¼šåŸå§‹æ•°æ®çš„æœ€å¤§å€¼å‡å»è¿‡æ»¤åæ•°æ®çš„ä¸­ä½æ•°
            delta1 <- max(data[, i], na.rm = TRUE) - median_value
            # å°†è®¡ç®—å¾—åˆ°çš„ delta1 å­˜å‚¨åœ¨ç»“æœå‘é‡ä¸­
            # i-1 æ˜¯å› ä¸ºæˆ‘ä»¬çš„ç»“æœå‘é‡ä»1å¼€å§‹ï¼Œè€Œå¾ªç¯ä»2å¼€å§‹
            delta1_values[i - 1] <- delta1
          } else {
            delta1_values[i - 1] <- NA
          }
        }
        # è¾“å‡ºdelta1å€¼
        delta1_values
        # ç¬¬äºŒéƒ¨åˆ†ï¼šè®¡ç®— Î»2 (delta2)ï¼ŒåŠ å…¥é˜ˆå€¼åˆ¤æ–­
        delta2_values <- numeric(ncol(data) - 1)
        for (i in 2:ncol(data)) {
          # æå–å½“å‰åˆ—çš„æ•°æ®å¹¶å»æ‰ç¼ºå¤±å€¼
          column_data <- na.omit(data[, i])
          
          # è®¡ç®—æœ€å¤§å€¼å’Œæ¬¡æœ€å¤§å€¼
          sorted_values <- sort(unique(column_data), decreasing = TRUE)
          max_value <- sorted_values[1]
          second_max_value <- ifelse(length(sorted_values) > 1, sorted_values[2], sorted_values[1])
          
          # æ£€æŸ¥æœ€å¤§å€¼ä¸æ¬¡æœ€å¤§å€¼çš„å·®æ˜¯å¦å¤§äºç­‰äº0.00001
          if (max_value - second_max_value >= 0.00001) {
            # åªå»æ‰æ¬¡æœ€å¤§å€¼
            column_data <- column_data[column_data != second_max_value]
            
            # æ£€æŸ¥å‰©ä½™æ•°æ®æ˜¯å¦è¶³å¤Ÿè®¡ç®—ä¸­ä½æ•°
            if (length(column_data) > 0) {
              median_value <- median(column_data)
              delta2 <- second_max_value - median_value
              delta2_values[i - 1] <- delta2
            } else {
              delta2_values[i - 1] <- NA
            }
          } else {
            # å¦‚æœå·®å€¼ä¸å¤§äº0.00001ï¼Œå­˜å‚¨NA
            delta2_values[i - 1] <- NA
          }
        }
        # è¾“å‡ºdelta2å€¼
        delta2_values
        # æ£€æŸ¥é•¿åº¦åŒ¹é…å¹¶æ›´æ–°max_data
        length_delta1 <- length(delta1_values)
        length_delta2 <- length(delta2_values)
        num_rows_max_data <- nrow(max_data)
        
        # æ·»åŠ é”™è¯¯å¤„ç†å’Œè¾“å‡ºé•¿åº¦ä¿¡æ¯
        print(paste("delta1_values length:", length_delta1))
        print(paste("delta2_values length:", length_delta2))
        print(paste("max_data rows:", num_rows_max_data))
        
        if (length_delta1 == num_rows_max_data && length_delta2 == num_rows_max_data) {
          # å°†deltaå€¼æ·»åŠ åˆ°max_dataä¸­
          max_data$Delta1_Value <- delta1_values
          max_data$Delta2_Value <- delta2_values
          
          # é€‰æ‹©éœ€è¦çš„åˆ—å¹¶åˆ›å»ºæ–°çš„æ•°æ®æ¡†
          result_data <- max_data[, c("Cluster", "Max_Cell_Cluster", "Delta1_Value", 
                                      "Second_Max_Cell_Cluster", "Delta2_Value")]
          # å°†ç»“æœå†™å…¥ä¸€ä¸ªæ–°çš„CSVæ–‡ä»¶
          write.csv(result_data, file = file.path(temp_folder, "result_updated.csv"), row.names = FALSE)
          result_updated0<-read.csv(file.path(temp_folder, "result_updated.csv"))
          
          output$resultTable <- renderTable({
            result_updated0[] <- lapply(result_updated0, function(x) if(is.numeric(x)) format(x, scientific = FALSE) else x)
            result_updated0
          })                   
          
            
          
          
        } else {
          cat("æ–‡ä»¶å¤åˆ¶å¤±è´¥ï¼Œè¯·æ£€æŸ¥ã€‚\n")
        }
      }
    })
    incProgress(0.8)
    incProgress(1 / total_steps)
    
    
    output$downloadData <- downloadHandler(
      filename = function() {
        paste("result_updated", Sys.Date(), ".csv", sep = "")
      },
      content = function(file) {
        # ç¡®ä¿æ–‡ä»¶è·¯å¾„æ­£ç¡®
        temp_file <- file.path(temp_folder, "result_updated.csv")
        if (file.exists(temp_file)) {
          file.copy(temp_file, file)
        } else {
          stop("The result file does not exist.")
        }
      }
    )
    
    # è®¡ç®—å®Œæˆåå…³é—­è¿›åº¦æ¡ï¼Œè¾“å‡ºâ€œè®¡ç®—å®Œæˆâ€
    shinyjs::alert("Calculation Finished!")
  })
  
  
  # è®¾ç½®ä¸Šä¼ æ–‡ä»¶å¤§å°é™åˆ¶
  options(shiny.maxRequestSize = 100 * 1024^2)  # 100 MB
  
  # æ¯ä¸ªæ ‡ç­¾é¡µçš„ååº”å¼å€¼
  umap_values <- reactiveValues(seurat_obj = NULL)
  heatmap_values <- reactiveValues(singleR_pred = NULL)
  delta_values <- reactiveValues(singleR_pred = NULL)
  
  # UMAPæ ‡ç­¾é¡µæ•°æ®å¤„ç†
  process_umap_data <- eventReactive(input$run_umap_analysis, {
    req(input$h5_file)  # ç¡®ä¿æ–‡ä»¶ä¸Šä¼ åæ‰å¤„ç†æ•°æ®
    
    withProgress(message = 'Processing data...', value = 0, {
      # Read data
      incProgress(0.1, detail = "Reading H5 file")
      hdf5_obj <- Read10X_h5(filename = input$h5_file$datapath, use.names = TRUE, unique.features = TRUE)
      
      # Create Seurat object and preprocess
      incProgress(0.2, detail = "Creating Seurat object")
      pbmc.seurat <- CreateSeuratObject(counts = hdf5_obj)
      
      # QC and filtering
      incProgress(0.3, detail = "QC and filtering")
      pbmc.seurat$mitoPercent <- PercentageFeatureSet(pbmc.seurat, pattern = '^MT-')
      pbmc.seurat.filtered <- subset(pbmc.seurat, 
                                     subset = nCount_RNA > 800 &
                                       nFeature_RNA > 500 &
                                       mitoPercent < 10)
      
      # Normalization and dimensionality reduction
      incProgress(0.4, detail = "Normalizing data")
      pbmc.seurat.filtered <- NormalizeData(pbmc.seurat.filtered)
      pbmc.seurat.filtered <- FindVariableFeatures(pbmc.seurat.filtered, selection.method = "vst", nfeatures = 2000)
      
      incProgress(0.5, detail = "Scaling data")
      pbmc.seurat.filtered <- ScaleData(pbmc.seurat.filtered, features = rownames(pbmc.seurat.filtered))
      
      incProgress(0.6, detail = "Running PCA")
      pbmc.seurat.filtered <- RunPCA(pbmc.seurat.filtered, features = VariableFeatures(pbmc.seurat.filtered))
      
      incProgress(0.7, detail = "Finding neighbors")
      pbmc.seurat.filtered <- FindNeighbors(pbmc.seurat.filtered, dims = 1:20)
      pbmc.seurat.filtered <- FindClusters(pbmc.seurat.filtered, resolution = 0.5)
      
      incProgress(0.8, detail = "Running UMAP")
      pbmc.seurat.filtered <- RunUMAP(pbmc.seurat.filtered, dims = 1:20)
      
      # Get reference data and run SingleR
      incProgress(0.9, detail = "Running SingleR")
      ref <- celldex::HumanPrimaryCellAtlasData()
      ref <- ref[,grepl('DC|B_cell|Neutrophils|T_cells|Monocyte|Erythroblast|Macrophage|NK_cell|Platelets|Myelocyte', ref$label.main)]
      
      pbmc_counts <- GetAssayData(pbmc.seurat.filtered, layer = 'counts')
      pred <- SingleR(test = pbmc_counts, ref = ref, labels = ref$label.main)
      
      # Add SingleR labels to Seurat object
      pbmc.seurat.filtered$singleR.labels <- pred$labels[match(rownames(pbmc.seurat.filtered@meta.data), rownames(pred))]
      
      # Store results
      umap_values$seurat_obj <- pbmc.seurat.filtered
      incProgress(1, detail = "Done!")
    })
  })
  
  # çƒ­å›¾æ ‡ç­¾é¡µæ•°æ®å¤„ç†
  process_heatmap_data <- eventReactive(input$run_heatmap_analysis, {
    req(input$h5_file)  # ç¡®ä¿æ–‡ä»¶ä¸Šä¼ åæ‰å¤„ç†æ•°æ®
    
    withProgress(message = 'Processing data...', value = 0, {
      # Read data
      incProgress(0.1, detail = "Reading H5 file")
      hdf5_obj <- Read10X_h5(filename = input$h5_file$datapath, use.names = TRUE, unique.features = TRUE)
      
      # Create Seurat object and preprocess
      incProgress(0.2, detail = "Creating Seurat object")
      pbmc.seurat <- CreateSeuratObject(counts = hdf5_obj)
      
      # QC and filtering
      incProgress(0.3, detail = "QC and filtering")
      pbmc.seurat$mitoPercent <- PercentageFeatureSet(pbmc.seurat, pattern = '^MT-')
      pbmc.seurat.filtered <- subset(pbmc.seurat, 
                                     subset = nCount_RNA > 800 &
                                       nFeature_RNA > 500 &
                                       mitoPercent < 10)
      
      # Normalization
      incProgress(0.4, detail = "Normalizing data")
      pbmc.seurat.filtered <- NormalizeData(pbmc.seurat.filtered)
      
      # Get reference data and run SingleR
      incProgress(0.8, detail = "Running SingleR")
      ref <- celldex::HumanPrimaryCellAtlasData()
      ref <- ref[,grepl('DC|B_cell|Neutrophils|T_cells|Monocyte|Erythroblast|Macrophage|NK_cell|Platelets|Myelocyte', ref$label.main)]
      
      pbmc_counts <- GetAssayData(pbmc.seurat.filtered, layer = 'counts')
      pred <- SingleR(test = pbmc_counts, ref = ref, labels = ref$label.main)
      
      # Store results
      heatmap_values$singleR_pred <- pred
      incProgress(1, detail = "Done!")
    })
  })
  
  # Deltaæ ‡ç­¾é¡µæ•°æ®å¤„ç†
  process_delta_data <- eventReactive(input$run_delta_analysis, {
    req(input$h5_file)  # ç¡®ä¿æ–‡ä»¶ä¸Šä¼ åæ‰å¤„ç†æ•°æ®
    
    withProgress(message = 'Processing data...', value = 0, {
      # Read data
      incProgress(0.1, detail = "Reading H5 file")
      hdf5_obj <- Read10X_h5(filename = input$h5_file$datapath, use.names = TRUE, unique.features = TRUE)
      
      # Create Seurat object and preprocess
      incProgress(0.2, detail = "Creating Seurat object")
      pbmc.seurat <- CreateSeuratObject(counts = hdf5_obj)
      
      # QC and filtering
      incProgress(0.3, detail = "QC and filtering")
      pbmc.seurat$mitoPercent <- PercentageFeatureSet(pbmc.seurat, pattern = '^MT-')
      pbmc.seurat.filtered <- subset(pbmc.seurat, 
                                     subset = nCount_RNA > 800 &
                                       nFeature_RNA > 500 &
                                       mitoPercent < 10)
      
      # Normalization
      incProgress(0.4, detail = "Normalizing data")
      pbmc.seurat.filtered <- NormalizeData(pbmc.seurat.filtered)
      
      # Get reference data and run SingleR
      incProgress(0.8, detail = "Running SingleR")
      ref <- celldex::HumanPrimaryCellAtlasData()
      ref <- ref[,grepl('DC|B_cell|Neutrophils|T_cells|Monocyte|Erythroblast|Macrophage|NK_cell|Platelets|Myelocyte', ref$label.main)]
      
      pbmc_counts <- GetAssayData(pbmc.seurat.filtered, layer = 'counts')
      pred <- SingleR(test = pbmc_counts, ref = ref, labels = ref$label.main)
      
      # Store results
      delta_values$singleR_pred <- pred
      incProgress(1, detail = "Done!")
    })
  })
  
  # ç”ŸæˆUMAPå›¾
  output$umap_plot <- renderPlot({
    req(umap_values$seurat_obj)
    DimPlot(
      umap_values$seurat_obj,
      reduction = 'umap',
      group.by = 'singleR.labels',
      label = TRUE,          # æ˜¾ç¤º cluster æ ‡ç­¾
      label.size = 6,        # å¢å¤§ cluster æ ‡ç­¾å­—ä½“ï¼ˆé»˜è®¤ 4ï¼‰
      repel = TRUE           # é˜²æ­¢æ ‡ç­¾é‡å 
    ) +
      ggtitle("UMAP visualization with SingleR annotations") +
      labs(x = "UMAP-1", y = "UMAP-2") +  # è‡ªå®šä¹‰åæ ‡è½´åç§°
      theme_classic() +
      theme(
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),  # ä¸»æ ‡é¢˜å±…ä¸­
        axis.title = element_text(size = 16, face = "bold"),  # åæ ‡è½´æ ‡é¢˜
        axis.text = element_text(size = 16),                 # åæ ‡è½´åˆ»åº¦æ•°å­—
        legend.title = element_text(size = 14),              # å›¾ä¾‹æ ‡é¢˜
        legend.text = element_text(size = 12)                 # å›¾ä¾‹å†…å®¹
      )
  })
  
  # ç”Ÿæˆå¾—åˆ†çƒ­å›¾
  output$score_heatmap <- renderPlot({
    req(heatmap_values$singleR_pred)
    # ç›´æ¥ä¿®æ”¹ plotScoreHeatmap å‡½æ•°è°ƒç”¨ï¼Œå¦‚æœå‡½æ•°æ¥å—å­—ä½“å¤§å°å‚æ•°
    plotScoreHeatmap(
      heatmap_values$singleR_pred,
      fontsize = 14,           # å¦‚æœå‡½æ•°æœ‰è¿™ä¸ªå‚æ•°
      fontsize_row = 14,       # è¡Œæ ‡ç­¾å­—ä½“å¤§å°
      fontsize_col = 14,       # åˆ—æ ‡ç­¾å­—ä½“å¤§å°
      main = "Score Heatmap",  # è®¾ç½®æ ‡é¢˜
      cexRow = 1.5,            # è¡Œæ ‡ç­¾å­—ä½“å¤§å°ç³»æ•°
      cexCol = 1.5             # åˆ—æ ‡ç­¾å­—ä½“å¤§å°ç³»æ•°
    )
  })
  
  # ç”ŸæˆDeltaåˆ†å¸ƒå›¾
  output$delta_plot <- renderPlot({
    req(delta_values$singleR_pred)
    # å¦‚æœ plotDeltaDistribution æ˜¯åŸºäº ggplot2 çš„
    p <- plotDeltaDistribution(delta_values$singleR_pred)
    p + theme(
      axis.title = element_text(size = 16, face = "bold"),
      axis.text = element_text(size = 16),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
      legend.title = element_text(size = 16),
      legend.text = element_text(size = 14),
      strip.text = element_text(size = 14, face = "bold"),  # åˆ†é¢æ ‡ç­¾æ–‡æœ¬
      panel.grid.major = element_line(size = 0.5),          # å¢åŠ ç½‘æ ¼çº¿ç²—ç»†
      panel.grid.minor = element_line(size = 0.25)
    ) +
      labs(x = "Delta Value", y = "Delta med") +
      guides(color = guide_legend(override.aes = list(size = 3)))  # å¢åŠ å›¾ä¾‹ç‚¹å¤§å°
  })
  
  # UMAPå›¾ä¸‹è½½åŠŸèƒ½
  output$download_umap <- downloadHandler(
    filename = function() {
      paste("umap-plot-", Sys.Date(), ".png", sep = "")
    },
    content = function(file) {
      req(umap_values$seurat_obj)
      
      # åˆ›å»ºUMAPå›¾
      p <- DimPlot(umap_values$seurat_obj, 
                   reduction = 'umap', 
                   group.by = 'singleR.labels',
                   label = TRUE,
                   label.size = 6,
                   repel = TRUE) +
        theme_classic() +
        labs(x = "UMAP-1", y = "UMAP-2") +
        ggtitle("UMAP visualization with SingleR annotations")+
        theme(
          plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
          axis.title = element_text(size = 16, face = "bold"),
          axis.text = element_text(size = 16),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12)
        )
      # ä½¿ç”¨ggsaveä¿å­˜å›¾åƒ
      ggsave(file, plot = p, width = 10, height = 8, dpi = 300)
    }
  )
  
  # çƒ­å›¾ä¸‹è½½åŠŸèƒ½
  output$download_heatmap <- downloadHandler(
    filename = function() {
      paste("score-heatmap-", Sys.Date(), ".png", sep = "")
    },
    content = function(file) {
      req(heatmap_values$singleR_pred)
      
      # åˆ›å»ºçƒ­å›¾å›¾
      p_heatmap <- plotScoreHeatmap(
        heatmap_values$singleR_pred,
        fontsize = 14,
        fontsize_row = 14,
        fontsize_col = 14,
        cexRow = 1.5,
        cexCol = 1.5,
        main = "Score Heatmap"
      )
      
      # ä½¿ç”¨ggsaveä¿å­˜å›¾åƒ
      ggsave(file, plot = p_heatmap, width = 10, height = 8, dpi = 300)
    }
  )
  # Deltaå›¾ä¸‹è½½åŠŸèƒ½
  output$download_delta <- downloadHandler(
    filename = function() {
      paste("delta-distribution-", Sys.Date(), ".png", sep = "")
    },
    content = function(file) {
      req(delta_values$singleR_pred)
      # æ„é€ å›¾å¯¹è±¡
      p <- plotDeltaDistribution(delta_values$singleR_pred) +
        theme(
          axis.title = element_text(size = 16, face = "bold"),
          axis.text = element_text(size = 16),
          plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
          legend.title = element_text(size = 16),
          legend.text = element_text(size = 14),
          strip.text = element_text(size = 14, face = "bold"),
          panel.grid.major = element_line(size = 0.5),
          panel.grid.minor = element_line(size = 0.25)
        ) +
        labs(x = "Delta Value", y = "Density") +
        guides(color = guide_legend(override.aes = list(size = 3)))
      
      # ä¿å­˜å›¾
      ggsave(file, plot = p, width = 10, height = 8, dpi = 300)
    }
  )
      
  
  # è§¦å‘æ•°æ®å¤„ç†
  observeEvent(input$run_umap_analysis, {
    process_umap_data()
  })
  
  observeEvent(input$run_heatmap_analysis, {
    process_heatmap_data()
  })
  
  observeEvent(input$run_delta_analysis, {
    process_delta_data()
  })
}



shinyApp(ui, server)